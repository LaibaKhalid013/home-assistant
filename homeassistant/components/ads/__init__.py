"""Support for Automation Device Specification (ADS)."""

import asyncio
from asyncio import timeout
from collections import namedtuple
import ctypes
from enum import Enum
import logging
import struct
import threading
from typing import Any

import pyads
import pyads.errorcodes
import voluptuous as vol

from homeassistant.const import (
    CONF_DEVICE,
    CONF_IP_ADDRESS,
    CONF_PORT,
    EVENT_HOMEASSISTANT_STOP,
)
from homeassistant.core import HomeAssistant, ServiceCall
import homeassistant.helpers.config_validation as cv
from homeassistant.helpers.entity import Entity
from homeassistant.helpers.typing import ConfigType

_LOGGER = logging.getLogger(__name__)

DATA_ADS = "data_ads"

# Supported Types
ADSTYPE_BOOL = "bool"
ADSTYPE_BYTE = "byte"
ADSTYPE_DINT = "dint"
ADSTYPE_INT = "int"
ADSTYPE_UDINT = "udint"
ADSTYPE_UINT = "uint"

ADS_TYPEMAP = {
    ADSTYPE_BOOL: pyads.PLCTYPE_BOOL,
    ADSTYPE_BYTE: pyads.PLCTYPE_BYTE,
    ADSTYPE_DINT: pyads.PLCTYPE_DINT,
    ADSTYPE_INT: pyads.PLCTYPE_INT,
    ADSTYPE_UDINT: pyads.PLCTYPE_UDINT,
    ADSTYPE_UINT: pyads.PLCTYPE_UINT,
}

CONF_ADS_FACTOR = "factor"
CONF_ADS_TYPE = "adstype"
CONF_ADS_VALUE = "value"
CONF_ADS_VAR = "adsvar"
CONF_ADS_VAR_BRIGHTNESS = "adsvar_brightness"
CONF_ADS_VAR_POSITION = "adsvar_position"

STATE_KEY_STATE = "state"
STATE_KEY_BRIGHTNESS = "brightness"
STATE_KEY_POSITION = "position"

DOMAIN = "ads"

SERVICE_WRITE_DATA_BY_NAME = "write_data_by_name"

CONFIG_SCHEMA = vol.Schema(
    {
        DOMAIN: vol.Schema(
            {
                vol.Required(CONF_DEVICE): cv.string,
                vol.Required(CONF_PORT): cv.port,
                vol.Optional(CONF_IP_ADDRESS): cv.string,
            }
        )
    },
    extra=vol.ALLOW_EXTRA,
)

SCHEMA_SERVICE_WRITE_DATA_BY_NAME = vol.Schema(
    {
        vol.Required(CONF_ADS_TYPE): vol.In(
            [
                ADSTYPE_INT,
                ADSTYPE_UINT,
                ADSTYPE_BYTE,
                ADSTYPE_BOOL,
                ADSTYPE_DINT,
                ADSTYPE_UDINT,
            ]
        ),
        vol.Required(CONF_ADS_VALUE): vol.Coerce(int),
        vol.Required(CONF_ADS_VAR): cv.string,
    }
)


async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
    """Set up the ADS component."""

    conf = config[DOMAIN]

    net_id = conf[CONF_DEVICE]
    ip_address = conf.get(CONF_IP_ADDRESS)
    port = conf[CONF_PORT]

    client = pyads.Connection(net_id, port, ip_address)

    try:
        ads = AdsHub(client)
    except pyads.ADSError:
        _LOGGER.error(
            "Could not connect to ADS host (netid=%s, ip=%s, port=%s)",
            net_id,
            ip_address,
            port,
        )
        return False

    hass.data[DATA_ADS] = ads
    hass.async_create_background_task(ads.heartbeat(), "heartbeat")
    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, ads.shutdown)

    def handle_write_data_by_name(call: ServiceCall) -> None:
        """Write a value to the connected ADS device."""
        ads_var = call.data[CONF_ADS_VAR]
        ads_type = call.data[CONF_ADS_TYPE]
        value = call.data[CONF_ADS_VALUE]

        try:
            ads.write_by_name(ads_var, value, ADS_TYPEMAP[ads_type])
        except pyads.ADSError as err:
            _LOGGER.error(err)

    hass.services.async_register(
        DOMAIN,
        SERVICE_WRITE_DATA_BY_NAME,
        handle_write_data_by_name,
        schema=SCHEMA_SERVICE_WRITE_DATA_BY_NAME,
    )

    return True


# Tuple to hold data needed for notification
NotificationItem = namedtuple(
    "NotificationItem", "hnotify huser name plc_datatype callback"
)

# Tuple to hold data needed to restore notification
DeletedNotification = namedtuple("DeletedNotification", "name plc_datatype callback")


class ConnectionState(Enum):
    """Reresentation of ADS connection state."""

    Connected = 1
    ReadyToReconnect = 2
    Disconnected = 3


class AdsHub:
    """Representation of an ADS connection."""

    def __init__(self, ads_client: pyads.Connection) -> None:
        """Initialize the ADS hub."""
        self._client = ads_client
        self._client.open()
        self._is_running = True

        # All ADS devices are registered here
        self._notification_items: dict[int, NotificationItem] = {}
        self._lock = threading.Lock()

    def shutdown(self, *args, **kwargs):
        """Shutdown ADS connection."""

        _LOGGER.debug("Shutting down ADS")

        self._is_running = False
        self._delete_device_notifications()

        try:
            self._client.close()
        except pyads.ADSError as err:
            _LOGGER.error(err)

    def write_by_name(self, name, value, plc_datatype):
        """Write a value to the device."""

        with self._lock:
            try:
                return self._client.write_by_name(name, value, plc_datatype)
            except pyads.ADSError as err:
                _LOGGER.error("Error writing %s: %s", name, err)

    def read_by_name(self, name, plc_datatype):
        """Read a value from the device."""

        with self._lock:
            try:
                return self._client.read_by_name(name, plc_datatype)
            except pyads.ADSError as err:
                _LOGGER.error("Error reading %s: %s", name, err)

    def add_device_notification(self, name, plc_datatype, callback):
        """Add a notification to the ADS devices."""

        attr = pyads.NotificationAttrib(ctypes.sizeof(plc_datatype))

        with self._lock:
            try:
                hnotify, huser = self._client.add_device_notification(
                    name, attr, self._device_notification_callback
                )
            except pyads.ADSError as err:
                _LOGGER.error("Error subscribing to %s: %s", name, err)
            else:
                hnotify = int(hnotify)
                self._notification_items[hnotify] = NotificationItem(
                    hnotify, huser, name, plc_datatype, callback
                )

                _LOGGER.debug(
                    "Added device notification %d for variable %s", hnotify, name
                )

    async def heartbeat(
        self,
        heartbeat_interval: float = 5.0,
        max_wait_time: float = 120.0,
        communication_timeout_while_disconnected_ms: int = 100,
    ):
        """Periodically checks and handles the connection state of the client.

        Adjusts the wait time and communication timeout based on the connection state, with exponential backoff when disconnected.

        Attempts reconnection and manages device notifications upon reconnection.
        """
        deleted_device_notifications: list[DeletedNotification] = []

        default_timeout_ms = 5000

        wait_time = heartbeat_interval

        while self._is_running:
            connection_state = self._check_connection()
            if connection_state == ConnectionState.Connected:
                if wait_time != heartbeat_interval:
                    _LOGGER.info("Reconnected")
                    self._client.set_timeout(ms=default_timeout_ms)
                    wait_time = heartbeat_interval

                await asyncio.sleep(wait_time)
            elif connection_state == ConnectionState.Disconnected:
                if wait_time == heartbeat_interval:
                    _LOGGER.info("Disconnected, waiting for device availability")
                    self._client.set_timeout(
                        ms=communication_timeout_while_disconnected_ms
                    )

                wait_time = (
                    wait_time * 2 if wait_time * 2 < max_wait_time else max_wait_time
                )
                _LOGGER.debug("Waiting %d seconds to check device state", wait_time)
                await asyncio.sleep(wait_time)
            elif connection_state == ConnectionState.ReadyToReconnect:
                if not deleted_device_notifications:
                    deleted_device_notifications = self._delete_device_notifications()

                self._reconnect()

                while deleted_device_notifications:
                    self.add_device_notification(*deleted_device_notifications.pop())

    def _check_connection(self) -> ConnectionState:
        try:
            self._client.read_state()
        except pyads.ADSError as read_state_error:
            if read_state_error.err_code not in pyads.errorcodes.ERROR_CODES:
                return ConnectionState.ReadyToReconnect
            return ConnectionState.Disconnected
        else:
            return ConnectionState.Connected

    def _reconnect(self):
        try:
            self._client.close()
            self._client.open()
        except pyads.ADSError as reconnect_error:
            _LOGGER.error(reconnect_error)
            return

    def _device_notification_callback(self, notification, name):
        """Handle device notifications."""
        contents = notification.contents

        hnotify = int(contents.hNotification)
        _LOGGER.debug("Received notification %d", hnotify)

        # get dynamically sized data array
        data_size = contents.cbSampleSize
        data = (ctypes.c_ubyte * data_size).from_address(
            ctypes.addressof(contents)
            + pyads.structs.SAdsNotificationHeader.data.offset
        )

        try:
            with self._lock:
                notification_item = self._notification_items[hnotify]
        except KeyError:
            _LOGGER.error("Unknown device notification handle: %d", hnotify)
            return

        # Parse data to desired datatype
        if notification_item.plc_datatype == pyads.PLCTYPE_BOOL:
            value = bool(struct.unpack("<?", bytearray(data))[0])
        elif notification_item.plc_datatype == pyads.PLCTYPE_INT:
            value = struct.unpack("<h", bytearray(data))[0]
        elif notification_item.plc_datatype == pyads.PLCTYPE_BYTE:
            value = struct.unpack("<B", bytearray(data))[0]
        elif notification_item.plc_datatype == pyads.PLCTYPE_UINT:
            value = struct.unpack("<H", bytearray(data))[0]
        elif notification_item.plc_datatype == pyads.PLCTYPE_DINT:
            value = struct.unpack("<i", bytearray(data))[0]
        elif notification_item.plc_datatype == pyads.PLCTYPE_UDINT:
            value = struct.unpack("<I", bytearray(data))[0]
        else:
            value = bytearray(data)
            _LOGGER.warning("No callback available for this datatype")

        notification_item.callback(notification_item.name, value)

    def _delete_device_notifications(self) -> list[DeletedNotification]:
        result = [
            DeletedNotification(item.name, item.plc_datatype, item.callback)
            for item in self._notification_items.values()
        ]

        while self._notification_items:
            _, notification_item = self._notification_items.popitem()
            _LOGGER.debug(
                "Deleting device notification %d, %d",
                notification_item.hnotify,
                notification_item.huser,
            )
            try:
                self._client.del_device_notification(
                    notification_item.hnotify, notification_item.huser
                )
            except pyads.ADSError as err:
                _LOGGER.error(err)

        return result


class AdsEntity(Entity):
    """Representation of ADS entity."""

    _attr_should_poll = False

    def __init__(self, ads_hub, name, ads_var) -> None:
        """Initialize ADS binary sensor."""
        self._state_dict: dict[str, Any] = {}
        self._state_dict[STATE_KEY_STATE] = None
        self._ads_hub = ads_hub
        self._ads_var = ads_var
        self._event = None
        self._attr_unique_id = ads_var
        self._attr_name = name

    async def async_initialize_device(
        self, ads_var, plctype, state_key=STATE_KEY_STATE, factor=None
    ):
        """Register device notification."""

        def update(name, value):
            """Handle device notifications."""
            _LOGGER.debug("Variable %s changed its value to %d", name, value)

            if factor is None:
                self._state_dict[state_key] = value
            else:
                self._state_dict[state_key] = value / factor

            asyncio.run_coroutine_threadsafe(async_event_set(), self.hass.loop)
            self.schedule_update_ha_state()

        async def async_event_set():
            """Set event in async context."""
            self._event.set()

        self._event = asyncio.Event()

        await self.hass.async_add_executor_job(
            self._ads_hub.add_device_notification, ads_var, plctype, update
        )
        try:
            async with timeout(10):
                await self._event.wait()
        except TimeoutError:
            _LOGGER.debug("Variable %s: Timeout during first update", ads_var)

    @property
    def available(self) -> bool:
        """Return False if state has not been updated yet."""
        return self._state_dict[STATE_KEY_STATE] is not None
