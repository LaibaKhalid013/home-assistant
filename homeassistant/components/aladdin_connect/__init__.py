"""The Aladdin Connect Genie integration."""

from __future__ import annotations

from typing import TYPE_CHECKING

from homeassistant.config_entries import ConfigEntry, ConfigEntryState
from homeassistant.helpers import issue_registry as ir

if TYPE_CHECKING:
    from homeassistant.core import HomeAssistant

DOMAIN = "aladdin_connect"


async def async_setup_entry(hass: HomeAssistant, _: ConfigEntry) -> bool:
    """Set up Aladdin Connect from a config entry."""
    ir.async_create_issue(
        hass,
        DOMAIN,
        DOMAIN,
        is_fixable=False,
        severity=ir.IssueSeverity.ERROR,
        translation_key="integration_removed",
        translation_placeholders={
            "entries": "/config/integrations/integration/aladdin_connect",
        },
    )

    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    if all(
        config_entry.state is ConfigEntryState.NOT_LOADED
        for config_entry in hass.config_entries.async_entries(DOMAIN)
        if config_entry.entry_id != entry.entry_id
    ):
        ir.async_delete_issue(hass, DOMAIN, DOMAIN)

    return True
