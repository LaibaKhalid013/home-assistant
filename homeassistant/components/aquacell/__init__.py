"""The Aquacell integration."""
from __future__ import annotations

from aioaquacell import AquacellApi

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import Platform
from homeassistant.core import HomeAssistant
from homeassistant.helpers.aiohttp_client import async_get_clientsession

from .const import DATA_AQUACELL, DOMAIN
from .coordinator import AquacellCoordinator

PLATFORMS = [Platform.SENSOR]


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up Aquacell from a config entry."""
    session = async_get_clientsession(hass)

    aquacell_api = AquacellApi(session)

    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {DATA_AQUACELL: aquacell_api}

    coordinator = AquacellCoordinator(hass, aquacell_api, entry)

    await coordinator.async_config_entry_first_refresh()
    hass.data[DOMAIN][entry.entry_id] = coordinator

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
        hass.data[DOMAIN].pop(entry.entry_id)

    return unload_ok
