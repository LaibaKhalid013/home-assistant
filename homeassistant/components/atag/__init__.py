"""The ATAG Integration."""

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import Platform
from homeassistant.core import HomeAssistant

from .coordinator import AtagDataUpdateCoordinator

DOMAIN = "atag"
PLATFORMS = [Platform.CLIMATE, Platform.SENSOR, Platform.WATER_HEATER]


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up Atag integration from a config entry."""

    coordinator = AtagDataUpdateCoordinator(hass, entry)
    await coordinator.async_config_entry_first_refresh()

    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
    if entry.unique_id is None:
        hass.config_entries.async_update_entry(entry, unique_id=coordinator.atag.id)

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload Atag config entry."""
    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)

    if unload_ok:
        hass.data[DOMAIN].pop(entry.entry_id)
    return unload_ok
