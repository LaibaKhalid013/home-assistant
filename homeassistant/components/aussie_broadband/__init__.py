"""The Aussie Broadband integration."""

from __future__ import annotations

from aiohttp import ClientError
from aussiebb.asyncio import AussieBB
from aussiebb.const import FETCH_TYPES
from aussiebb.exceptions import AuthenticationException

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_PASSWORD, CONF_USERNAME, Platform
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady
from homeassistant.helpers.aiohttp_client import async_get_clientsession

from .const import DOMAIN
from .coordinator import AussieBroadandDataUpdateCoordinator

PLATFORMS = [Platform.SENSOR]


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up Aussie Broadband from a config entry."""
    # Login to the Aussie Broadband API and retrieve the current service list
    client = AussieBB(
        entry.data[CONF_USERNAME],
        entry.data[CONF_PASSWORD],
        async_get_clientsession(hass),
    )

    # Ignore services that don't support usage data
    ignore_types = [*FETCH_TYPES, "Hardware"]

    try:
        await client.login()
        services = await client.get_services(drop_types=ignore_types)
    except AuthenticationException as exc:
        raise ConfigEntryAuthFailed from exc
    except ClientError as exc:
        raise ConfigEntryNotReady from exc

    # Initiate a Data Update Coordinator for each service
    for service in services:
        service["coordinator"] = AussieBroadandDataUpdateCoordinator(
            hass, client, service["service_id"]
        )
        await service["coordinator"].async_config_entry_first_refresh()

    # Setup the integration
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
        "client": client,
        "services": services,
    }
    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload the config entry."""
    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
    if unload_ok:
        hass.data[DOMAIN].pop(entry.entry_id)

    return unload_ok
