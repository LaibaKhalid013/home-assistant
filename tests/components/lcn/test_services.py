"""Test for the LCN services."""

from unittest.mock import patch

import pytest
from syrupy import SnapshotAssertion

from homeassistant.components.lcn import DOMAIN
from homeassistant.core import HomeAssistant
from homeassistant.setup import async_setup_component

from .conftest import MockModuleConnection, MockPchkConnectionManager, setup_component


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_output_abs(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test output_abs service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "dim_output") as dim_output:
        await hass.services.async_call(
            DOMAIN,
            "output_abs",
            {
                "address": "pchk.s0.m7",
                "output": "output1",
                "brightness": 100,
                "transition": 5,
            },
            blocking=True,
        )

    assert dim_output.await_args.args == snapshot(name="dim_output")


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_output_rel(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test output_rel service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "rel_output") as rel_output:
        await hass.services.async_call(
            DOMAIN,
            "output_rel",
            {
                "address": "pchk.s0.m7",
                "output": "output1",
                "brightness": 25,
            },
            blocking=True,
        )

    assert rel_output.await_args.args == snapshot(name="rel_output")


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_output_toggle(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test output_toggle service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "toggle_output") as toggle_output:
        await hass.services.async_call(
            DOMAIN,
            "output_toggle",
            {
                "address": "pchk.s0.m7",
                "output": "output1",
                "transition": 5,
            },
            blocking=True,
        )

    assert toggle_output.await_args.args == snapshot(name="toggle_output")


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_relays(hass: HomeAssistant, snapshot: SnapshotAssertion) -> None:
    """Test relays service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "control_relays") as control_relays:
        await hass.services.async_call(
            DOMAIN,
            "relays",
            {"address": "pchk.s0.m7", "state": "0011TT--"},
            blocking=True,
        )

    assert control_relays.await_args.args == snapshot(name="control_relays")


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_led(hass: HomeAssistant, snapshot: SnapshotAssertion) -> None:
    """Test led service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "control_led") as control_led:
        await hass.services.async_call(
            DOMAIN,
            "led",
            {"address": "pchk.s0.m7", "led": "led6", "state": "blink"},
            blocking=True,
        )

    assert control_led.await_args.args == snapshot(name="control_led")


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_var_abs(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test var_abs service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "var_abs") as var_abs:
        await hass.services.async_call(
            DOMAIN,
            "var_abs",
            {
                "address": "pchk.s0.m7",
                "variable": "var1",
                "value": 75,
                "unit_of_measurement": "%",
            },
            blocking=True,
        )

    assert var_abs.await_args.args == snapshot(name="var_abs")


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_var_rel(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test var_rel service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "var_rel") as var_rel:
        await hass.services.async_call(
            DOMAIN,
            "var_rel",
            {
                "address": "pchk.s0.m7",
                "variable": "var1",
                "value": 10,
                "unit_of_measurement": "%",
                "value_reference": "current",
            },
            blocking=True,
        )

    assert var_rel.await_args.args == snapshot(name="var_rel")


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_var_reset(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test var_reset service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "var_reset") as var_reset:
        await hass.services.async_call(
            DOMAIN,
            "var_reset",
            {"address": "pchk.s0.m7", "variable": "var1"},
            blocking=True,
        )

    assert var_reset.await_args.args == snapshot(name="var_reset")


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_lock_regulator(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test lock_regulator service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "lock_regulator") as lock_regulator:
        await hass.services.async_call(
            DOMAIN,
            "lock_regulator",
            {"address": "pchk.s0.m7", "setpoint": "r1varsetpoint", "state": True},
            blocking=True,
        )

    assert lock_regulator.await_args.args == snapshot(name="lock_regulator")


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_send_keys(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test send_keys service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "send_keys") as send_keys:
        await hass.services.async_call(
            DOMAIN,
            "send_keys",
            {"address": "pchk.s0.m7", "keys": "a1a5d8", "state": "hit"},
            blocking=True,
        )

    keys = [[False] * 8 for i in range(4)]
    keys[0][0] = True
    keys[0][4] = True
    keys[3][7] = True

    assert send_keys.await_args.args == snapshot(name="send_keys")


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_send_keys_hit_deferred(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test send_keys (hit_deferred) service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    keys = [[False] * 8 for i in range(4)]
    keys[0][0] = True
    keys[0][4] = True
    keys[3][7] = True

    # success
    with patch.object(
        MockModuleConnection, "send_keys_hit_deferred"
    ) as send_keys_hit_deferred:
        await hass.services.async_call(
            DOMAIN,
            "send_keys",
            {"address": "pchk.s0.m7", "keys": "a1a5d8", "time": 5, "time_unit": "s"},
            blocking=True,
        )

    assert send_keys_hit_deferred.await_args.args == snapshot(
        name="send_keys_hit_deferred"
    )

    # wrong key action
    with (
        patch.object(
            MockModuleConnection, "send_keys_hit_deferred"
        ) as send_keys_hit_deferred,
        pytest.raises(ValueError),
    ):
        await hass.services.async_call(
            DOMAIN,
            "send_keys",
            {
                "address": "pchk.s0.m7",
                "keys": "a1a5d8",
                "state": "make",
                "time": 5,
                "time_unit": "s",
            },
            blocking=True,
        )


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_lock_keys(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test lock_keys service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "lock_keys") as lock_keys:
        await hass.services.async_call(
            DOMAIN,
            "lock_keys",
            {"address": "pchk.s0.m7", "table": "a", "state": "0011TT--"},
            blocking=True,
        )

    assert lock_keys.await_args.args == snapshot(name="lock_keys")


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_lock_keys_tab_a_temporary(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test lock_keys (tab_a_temporary) service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    # success
    with patch.object(
        MockModuleConnection, "lock_keys_tab_a_temporary"
    ) as lock_keys_tab_a_temporary:
        await hass.services.async_call(
            DOMAIN,
            "lock_keys",
            {
                "address": "pchk.s0.m7",
                "state": "0011TT--",
                "time": 10,
                "time_unit": "s",
            },
            blocking=True,
        )

    assert lock_keys_tab_a_temporary.await_args.args == snapshot(
        name="lock_keys_tab_a_temporary"
    )

    # wrong table
    with (
        patch.object(
            MockModuleConnection, "lock_keys_tab_a_temporary"
        ) as lock_keys_tab_a_temporary,
        pytest.raises(ValueError),
    ):
        await hass.services.async_call(
            DOMAIN,
            "lock_keys",
            {
                "address": "pchk.s0.m7",
                "table": "b",
                "state": "0011TT--",
                "time": 10,
                "time_unit": "s",
            },
            blocking=True,
        )


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_dyn_text(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test dyn_text service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "dyn_text") as dyn_text:
        await hass.services.async_call(
            DOMAIN,
            "dyn_text",
            {"address": "pchk.s0.m7", "row": 1, "text": "text in row 1"},
            blocking=True,
        )

    assert dyn_text.await_args.args == snapshot(name="dyn_text")


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_pck(hass: HomeAssistant, snapshot: SnapshotAssertion) -> None:
    """Test pck service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "pck") as pck:
        await hass.services.async_call(
            DOMAIN,
            "pck",
            {"address": "pchk.s0.m7", "pck": "PIN4"},
            blocking=True,
        )

    assert pck.await_args.args == snapshot(name="pck")


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_called_with_invalid_host_id(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test service was called with non existing host id."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "pck") as pck, pytest.raises(ValueError):
        await hass.services.async_call(
            DOMAIN,
            "pck",
            {"address": "foobar.s0.m7", "pck": "PIN4"},
            blocking=True,
        )

    pck.assert_not_awaited()
