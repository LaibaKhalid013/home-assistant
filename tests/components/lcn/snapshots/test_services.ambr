# serializer version: 1
# name: test_service_dyn_text[dyn_text]
  tuple(
    0,
    'text in row 1',
  )
# ---
# name: test_service_led[led]
  tuple(
    <LedPort.LED6: 5>,
    <LedStatus.BLINK: 'B'>,
  )
# ---
# name: test_service_lock_keys[lock_keys]
  tuple(
    0,
    list([
      <KeyLockStateModifier.OFF: '0'>,
      <KeyLockStateModifier.OFF: '0'>,
      <KeyLockStateModifier.ON: '1'>,
      <KeyLockStateModifier.ON: '1'>,
      <KeyLockStateModifier.TOGGLE: 'U'>,
      <KeyLockStateModifier.TOGGLE: 'U'>,
      <KeyLockStateModifier.NOCHANGE: '-'>,
      <KeyLockStateModifier.NOCHANGE: '-'>,
    ]),
  )
# ---
# name: test_service_lock_keys_tab_a_temporary[lock_keys-tab-a-temporary-success]
  tuple(
    10,
    <TimeUnit.SECONDS: 'S'>,
    list([
      <KeyLockStateModifier.OFF: '0'>,
      <KeyLockStateModifier.OFF: '0'>,
      <KeyLockStateModifier.ON: '1'>,
      <KeyLockStateModifier.ON: '1'>,
      <KeyLockStateModifier.TOGGLE: 'U'>,
      <KeyLockStateModifier.TOGGLE: 'U'>,
      <KeyLockStateModifier.NOCHANGE: '-'>,
      <KeyLockStateModifier.NOCHANGE: '-'>,
    ]),
  )
# ---
# name: test_service_lock_regulator[lock_regulator]
  tuple(
    0,
    True,
  )
# ---
# name: test_service_output_abs[output_abs]
  tuple(
    0,
    100,
    9,
  )
# ---
# name: test_service_output_rel[output_rel]
  tuple(
    0,
    25,
  )
# ---
# name: test_service_output_toggle[output_toggle]
  tuple(
    0,
    9,
  )
# ---
# name: test_service_pck[pck]
  tuple(
    'PIN4',
  )
# ---
# name: test_service_relays[relays]
  tuple(
    list([
      <RelayStateModifier.OFF: '0'>,
      <RelayStateModifier.OFF: '0'>,
      <RelayStateModifier.ON: '1'>,
      <RelayStateModifier.ON: '1'>,
      <RelayStateModifier.TOGGLE: 'U'>,
      <RelayStateModifier.TOGGLE: 'U'>,
      <RelayStateModifier.NOCHANGE: '-'>,
      <RelayStateModifier.NOCHANGE: '-'>,
    ]),
  )
# ---
# name: test_service_send_keys[send_keys]
  tuple(
    list([
      list([
        True,
        False,
        False,
        False,
        True,
        False,
        False,
        False,
      ]),
      list([
        False,
        False,
        False,
        False,
        False,
        False,
        False,
        False,
      ]),
      list([
        False,
        False,
        False,
        False,
        False,
        False,
        False,
        False,
      ]),
      list([
        False,
        False,
        False,
        False,
        False,
        False,
        False,
        True,
      ]),
    ]),
    <SendKeyCommand.HIT: 'K'>,
  )
# ---
# name: test_service_send_keys_hit_deferred[send_keys-hit-deferred-success]
  tuple(
    list([
      list([
        True,
        False,
        False,
        False,
        True,
        False,
        False,
        False,
      ]),
      list([
        False,
        False,
        False,
        False,
        False,
        False,
        False,
        False,
      ]),
      list([
        False,
        False,
        False,
        False,
        False,
        False,
        False,
        False,
      ]),
      list([
        False,
        False,
        False,
        False,
        False,
        False,
        False,
        True,
      ]),
    ]),
    5,
    <TimeUnit.SECONDS: 'S'>,
  )
# ---
# name: test_service_var_abs[var_abs]
  tuple(
    <Var.VAR1ORTVAR: 0>,
    75.0,
    <VarUnit.PERCENT: '%'>,
  )
# ---
# name: test_service_var_rel[var_rel]
  tuple(
    <Var.VAR1ORTVAR: 0>,
    10.0,
    <VarUnit.PERCENT: '%'>,
    <RelVarRef.CURRENT: 1>,
  )
# ---
# name: test_service_var_reset[var_reset]
  tuple(
    <Var.VAR1ORTVAR: 0>,
  )
# ---
